Mathlib/Combinatorics/Flow/Basic.lean

```diff
@@ -0,0 +1,167 @@
+- `Network.Problem`: A `Network` along with fixed source and sink vertices $s$ and $t$.
+- `Flow`: A valid assignment of flow values for a given `Network.Problem`.
+
+## Notation
+
+For flows `F F' : Flow Pr`, we define two relations:
+
+ - `F ≤ F'`: Equivalent to `F.value ≤ F'.value`.
+ - `F ⊆ F'`: Equivalent to `F.f ≤ F'.f`.
+
+## References
+
+See [goldberg_network_1989] for a thorough introduction to the topic.
+
+## Tags
+
+flow, network, max-flow, maxflow, cut, min-cut, mincut, graph, subflow
+-/
+
+universe u_V u_R
+
+open BigOperators
+
+section NetworkDefinition
+
+variable (V : Type u_V) (R : Type u_R) [Zero R] [LE R]
+
+/-- A flow network assigning non-negative capacities to all pairs of vertices. -/
+structure Network where
+  /-- The assigned capacities. -/
+  cap : V → V → R
+  nonneg : ∀ u v, 0 ≤ cap u v
+  loopless: ∀ v, cap v v = 0
+
+/-- A `Network` where all capacities are symmetrical. -/
+structure UndirectedNetwork extends Network V R where
+  symm : ∀ u v, cap u v = cap v u
+
+end NetworkDefinition
+
+variable {V : Type u_V} {R : Type u_R}
+
+namespace UndirectedNetwork
+
+/-- The graph made of all edges that have positive capacity in the network. -/
+def asSimpleGraph [Zero R] [Preorder R] (N : UndirectedNetwork V R) : SimpleGraph V where
+  Adj u v := 0 < N.cap u v
+  symm u v huv := by simp only [N.symm, huv]
+  loopless v hv := hv.ne.symm <| N.loopless v
+
+end UndirectedNetwork
+
+/-- A fixed choice of _source_ and _sink_ vertices $s$ and $t$ in a given `Network`. -/
+structure Network.Problem [Zero R] [LE R] (N : Network V R) where
+  /-- The source vertex. -/
+  s : V
+  /-- The sink vertex. -/
+  t : V
+
+variable [Fintype V]
+
+section FlowDefinition
+
+/-- The incoming flow of a vertex. --/
+def flowIn [AddCommMonoid R] (f : V → V → R) (v : V) := ∑ u, f u v
+/-- The outgoing flow of a vertex. --/
+def flowOut [AddCommMonoid R] (f : V → V → R) (v : V) := ∑ w, f v w
+/-- The excess flow of a vertex (the incoming flow minus the outgoing flow). --/
+def excess [SubtractionCommMonoid R] (f : V → V → R) (v : V) := flowIn f v - flowOut f v
+
+variable [OrderedAddCommMonoid R] {N : Network V R}
+
+/-- A valid assignment of flow units to all edges for a given `Network.Problem`.
+
+    Note that this definition, in contrast to what can sometimes be found in the literature, does
+    not require that $f(u, v) = -f(v, u)$. This extra requirement eliminates cycles between two
+    vertices and makes some theorem statements shorter. However, since larger cycles are still
+    permitted, this usually does not help in proofs. To the contrary, defining a flow becomes more
+    cumbersome, sometimes significantly so. -/
+@[ext]
+structure Network.Problem.Flow (Pr : N.Problem) where
+  /-- The assignment of flow units to each edge. -/
+  f : V → V → R
+  nonneg : ∀ u v, 0 ≤ f u v
+  conservation : ∀ v, v ≠ Pr.s ∧ v ≠ Pr.t → flowOut f v = flowIn f v
+  capacity : ∀ u v, f u v ≤ N.cap u v
```

Comment:
I would suggest following the `FunLike` design
```suggestion
structure Network.Problem.Flow (Pr : N.Problem) where
  /-- The assignment of flow units to each edge.
  
  Do NOT use. Use the coercion to function instead. -/
  toFun : V → V → R
  nonneg' : ∀ u v, 0 ≤ toFun u v
  conservation' : ∀ v, v ≠ Pr.s ∧ v ≠ Pr.t → flowOut toFun v = flowIn toFun v
  capacity' : ∀ u v, toFun u v ≤ N.cap u v
  
 instance : FunLike Pr.Flow V (V → R) where
   coe := toFun
   coe_injective' f g h := by cases f; cases g; congr'
   
 -- Here reprove `nonneg`, `conservation`, `capacity` in terms of `⇑F` where `F : Pr.Flow`, rather than `F.toFun`
```
Please grep for bits of the code above to see what I am expecting